openmls/src/group/tests/test_proposal_validation.rs

```diff
@@ -656,92 +669,183 @@ fn test_valsem103(ciphersuite: Ciphersuite, backend: &impl OpenMlsCryptoProvider
+    ] {
+        // We now test if ValSem103 is also performed when a client receives a
+        // commit. Before we can test reception of (invalid) proposals, we set up a
+        // new group with Alice and Bob.
+        let ProposalValidationTestSetup {
+            mut alice_group,
+            mut bob_group,
+        } = validation_test_setup(PURE_PLAINTEXT_WIRE_FORMAT_POLICY, ciphersuite, backend);
+
+        // We now have alice create a commit. Then we artificially add an Add
+        // proposal with an existing identity (Bob).
+        let (_bob_credential_bundle, bob_key_package_bundle) =
+            generate_credential_bundle_and_key_package_bundle("Bob".into(), ciphersuite, backend);
+        let bob_key_package = bob_key_package_bundle.key_package().clone();
 
-    // Keep the original plaintext for positive test later.
-    let original_plaintext = plaintext.clone();
+        // Create the Commit.
+        let serialized_update = alice_group
+            .self_update(backend, None)
+            .expect("Error creating self-update")
+            .tls_serialize_detached()
+            .expect("Could not serialize message.");
 
-    let add_proposal = Proposal::Add(AddProposal {
-        key_package: bob_key_package,
-    });
+        let plaintext = MlsMessageIn::tls_deserialize(&mut serialized_update.as_slice())
+            .expect("Could not deserialize message.")
+            .into_plaintext()
+            .expect("Message was not a plaintext.");
+        // Keep the original plaintext for positive test later.
+        let original_plaintext = plaintext.clone();
 
-    // Artificially add a proposal trying to add (another) Bob.
-    let verifiable_plaintext = insert_proposal_and_resign(
-        backend,
-        ProposalOrRef::Proposal(add_proposal),
-        plaintext,
-        &original_plaintext,
-        &alice_group,
-    );
+        let proposals = match alice_and_bob_share_identities {
+            KeyUniqueness::NegativeSameKey => {
+                let add_proposal = ProposalOrRef::Proposal(Proposal::Add(AddProposal {
+                    key_package: bob_key_package,
+                }));
+                vec![add_proposal]
+            }
+            KeyUniqueness::PositiveSameKeyWithRemove | KeyUniqueness::PositiveDifferentKey => {
+                let add_proposal = ProposalOrRef::Proposal(Proposal::Add(AddProposal {
+                    key_package: bob_key_package,
+                }));
+                // find bob's index
+                let bob_index = alice_group
+                    .members()
+                    .find_map(|member| {
+                        if member.identity == b"Bob" {
+                            Some(member.index)
+                        } else {
+                            None
+                        }
+                    })
+                    .unwrap();
+                let remove_proposal = ProposalOrRef::Proposal(Proposal::Remove(RemoveProposal {
```

Comment:
Ok so for `PositiveDifferentKey` the commit will also have a remove proposal ? This is not right. Just make this step unreachable